### PR TITLE
coc-groovy gradle support

### PR DESCRIPTION
**Pull request by joshzcold**
- Check for pom.xml first then attempt to find build.gradle
- if build.gradle find if the user has the wrapper script `gradlew` or `gradle.bat` for windows. Generated by `gradle init` else use regular gradle command
- Handle if `groovy.project.referencedLibraries` turns out to be a colon seperated classpath string instead of an array.
- add gradle instructions to readme

biggest problem I have is finding a built in way to generate the classpath in gradle. Nothing built in like maven as far as I can tell. 

I have added this in the readme for users to add
```groovy
tasks.register("classPath"){
    def classpath = sourceSets.main.compileClasspath.asPath 
    if(System.getProperty("outputFile") != ""){
        File out =  new File(outputFile)
        out.getParentFile().mkdirs()
        out.text = classpath
    }
    else
        println classpath
}
```
However having to manually add something like this kinda goes against the plug-and-play spirit of coc extensions.
If we could think of a better way before merge that would be nice.

### CODE REVIEW

**Comment by dansomething** [comment]
I'll have time to look at the code later... That said, here are some initial thoughts based on your description.

> biggest problem I have is finding a built in way to generate the classpath in gradle. Nothing built in like maven as far as I can tell.

This is exactly why I didn't add the implementation myself. I'll have to think about it more. It's definitely not in the sprit of coc.nvim. What will happen if the Gradle build isn't setup properly? How will it handle this error? How will the user be presented with this problem? Like you said, can we find a better way than requiring a modification of the Gradle build?

> Handle if groovy.project.referencedLibraries turns out to be a colon seperated classpath string instead of an array.

Please explain how/why this would happen.

**Comment by dansomething** [comment]
relates to #11

**Comment by joshzcold** [comment]
> Handle if groovy.project.referencedLibraries turns out to be a colon seperated classpath string instead of an array.
> Please explain how/why this would happen.

so in `classpath.ts` https://github.com/dansomething/coc-groovy/blob/87e1b2df4e59aec29432d17c45ae366ab3ad903c/src/classpath.ts#L26

the config get call is of type `<string[]>` however surprisingly if you only have a string in the config section `referencedLibraries` then type script doesn't throw an error and still tries to send the concated classpath string to the server.

this concated string is sent to the server as type `JsonPrimitive` instead of what they are checking for `JsonArray` 
Because the class path is a string instead of an array send to the server it rejects the classpath and nothing gets resolved.

I figured rather than having scenario where no errors are thrown and nothing gets resolved it would be better to give the ability to the user to put in a long string they probably copied from another programs output.

Something like `myjar1.jar:myjar2.jar:etc:etc:etc`

However it would be just as good if coc-groovy threw an error if it turned out not to be an array.

**Comment by joshzcold** [comment]
Related to https://github.com/prominic/groovy-language-server/issues/26
As they are trying to figure out the best way to resolve a gradle classpath on their end.
do they resolve classpaths with maven? seems like they would leave it up to the clients to resolve

**Comment by joshzcold** [comment]
marking as a Work In Progress until there is something more plugin-and-play we can do. 
Until we think of something there is nothing stopping a user from pulling my fork via 
`Plug 'joshzcold/coc-groovy', {'do': 'yarn install --frozen-lockfile'}`
and adding that section to their `build.gradle` to get this working in the mean time.

**Comment by dansomething** [comment]
> I figured rather than having scenario where no errors are thrown and nothing gets resolved it would be better to give the ability to the user to put in a long string they probably copied from another programs output

This is fine. It stays with the coc philosophy of it "just works". I can see how some people might do a copy/paste mistake like that and spitting out an error isn't as friendly.

re: Generating the class path.
I just threw together a tool which might do the job. Would you mind trying it out?

It will attempt to externally generate the class path for a Gradle project. If it works reliably enough it can be used instead of requiring modifications to the build.gradle file. You may want to compare it's results to yours. This solution could output more libraries as it will include all scopes: compile, test, etc.

[gradle-classpath.zip](https://github.com/dansomething/coc-groovy/files/6411989/gradle-classpath.zip)
```sh
# Print classpath
$> cd gradle/project
$> /path/to/gradle-classpath/bin/gradle-classpath

# View Help
$> /path/to/gradle-classpath/bin/gradle-classpath -h
```

**Comment by joshzcold** [comment]
@dansomething  works great. Had to remove my classPath task in build.gradle, but the odds of a user having that is very slim. 

tried it with and without `-o`, no errors. `-o` also rewrites the file sucessfully.

File: src/classpath.ts
Comment (joshzcold) [inline src/classpath.ts]: You might want to test out this `echo %errorlevel%` on a windows machine. would effect both the mvn command and the gradle command.

**Comment by joshzcold** [comment]
changes I made utilizing gradle-classpath is working on my machine.

File: src/classpath.ts
Comment (joshzcold) [reply src/classpath.ts]: got it from this page: https://stackoverflow.com/questions/334879/how-do-i-get-the-application-exit-code-from-a-windows-command-line/11476681#11476681

Comment (dansomething) [reply src/classpath.ts]: What OS(s) have you tested on?

**Comment by dansomething** [comment]
Go ahead and fix the build issues and then I'll take a deeper look at the code.

File: src/classpath.ts
Comment (joshzcold) [reply src/classpath.ts]: im just on Linux (Arch)

File: src/classpath.ts
Comment (dansomething) [inline src/classpath.ts]: `isPom` seems inappropriate since it's a build file path. Let's use `let buildFile = ...` here. Then, reassign if it falls into the `if` condition for Gradle. After that there should just be one `return buildFile;` at the end of this method.

File: src/classpath.ts
Comment (dansomething) [inline src/classpath.ts]: Let's use the `--path-separator` argument here too. That way if the default ever changes it won't break this code and it will make the expectation clear.

File: src/classpath.ts
Comment (dansomething) [inline src/classpath.ts]: There are a couple of small issues here. `builtClassPath` could be `null` so we should conditionally concat it in both cases. With that, it should be possible to simplify this a bit doing the concat once and also alleviate the need to cast `builtClassPath`. Let's also stick with `as` for casting.

Here's what I'm thinking:
```javascript
  let classpath: Array<string> | string = config.get<string[]>(Settings.REFERENCED_LIBRARIES, []);
  if (!Array.isArray(classpath)) {
    const value = classpath as string;
    classpath = value.split(separator) as string[];
  }
  if (builtClassPath) {
    classpath = classpath.concat(builtClassPath);
  }
```

File: src/classpath.ts
Comment (dansomething) [inline src/classpath.ts]: Let's replace this with `let cmd: string;` instead of using the default `''` to improve clarity when no `cmd` is set.

File: src/classpath.ts
Comment (dansomething) [inline src/classpath.ts:73]: It's clear this method is getting a bit complex to read with, I think, 7 different returns. This is more of just a future reference to refactor this to clean it up. It doesn't need to be done now.

File: src/classpath.ts
Comment (dansomething) [inline src/classpath.ts]: We'll want to test this on Windows before merging. I can try to find some time later for that.

Comment (dansomething) [reply src/classpath.ts]: Did you test the failure case on Linux?

Comment (dansomething) [reply src/classpath.ts]: I could also easily update `gradle-classpath` to output 'BUILD SUCCESS' or some other success message and go back to checking the result if that's easier. Thoughts?

File: src/classpath.ts
Comment (dansomething) [inline src/classpath.ts]: I think we're going to want the extension path here. The `workspace.root` will be for the opened Groovy project. Also, `groovy-classpath` should be `gradle-classpath` and we'll want to use `path.resolve` to avoid hardcoding the path separator. I believe this method could simply become:

```javascript
  return path.resolve(
    extensions.getExtension('coc-groovy').extension.extensionPath,
    'utils',
    'gradle-classpath',
    'bin',
    'gradle-classpath',
    IS_WINDOWS ? '.bat' : ''
  );
```

When you test this, be sure to delete the coc-groovy-data directory for the project or use `:CocCommand groovy.project.config.update` to force it to be regenerated.

**Review by dansomething — CHANGES_REQUESTED** [review]
This is good progress. It will be great to have Gradle support. There are a couple things I'd like to address where I've left comments. After those, along with a some testing on Windows, it will hopefully be ready to go.

File: src/classpath.ts
Comment (joshzcold) [reply src/classpath.ts]: I think checking the exit code should be solid enough. More so than a string check.

File: src/classpath.ts
Comment (joshzcold) [reply src/classpath.ts]: woops didn't catch that `groovy-classpath` thanks. Must have been waking up or something.

File: src/classpath.ts
Comment (joshzcold) [reply src/classpath.ts]: @dansomething my current gradle-classpath jar is missing `--path-seperator` got a link or download?

Comment (dansomething) [reply src/classpath.ts]: Ah sorry about that, I must have added that right after I uploaded? Here you go:
~~[gradle-classpath.zip](https://github.com/dansomething/coc-groovy/files/6377782/gradle-classpath.zip)~~ See https://github.com/dansomething/coc-groovy/pull/12#issuecomment-825938296 for the latest version.

I also updated the other comment link. As soon as I find the time, I'll upload this project to GitHub.

**Comment by joshzcold** [comment]
```
`groovy.project.referencedLibraries: []` ✅
`groovy.project.referencedLibraries: ""` ✅
`groovy.project.referencedLibraries: "somethingthatDoesn'tExist"` ✅
`groovy.project.referencedLibraries: "somethingThatDoesn'tExist.jar:SometyhignElseThatDoesn'tExist.jar"` ✅
`"referencedLibraries":["build/libs/sspr-1.0.jar","gradle/wrapper/gradle-wrapper.jar" ]` ✅ // didn't throw an error

"referencedLibraries":["/home/joshua/git/sspr-automation/build/libs/sspr-1.0.jar",
"/home/joshua/git/sspr-automation/gradle/wrapper/gradle-wrapper.jar" ]
// was able to import org.gradle.wrapper.Logger ✅

"referencedLibraries":"/home/joshua/git/sspr-automation/build/libs/sspr-1.0.jar:/home/joshua/git/sspr-automation/gradle/wrapper/gradle-wrapper.jar"
// was able to import org.gradle.wrapper.Logger ✅

"referencedLibraries":"TryingtoBreakThis.jar;TryingToBreakThiswithASemiColon" ✅

"referencedLibraries":"TryingtoBreakThis.jar:::::::::::::::::::::"
// doesn't break ✅
```
1. update config
2. `CocRestart`
3. `CocCommand groovy.project.config.update`

File: src/classpath.ts
Comment (joshzcold) [reply src/classpath.ts]: I tested pass and failure cases of editing the configuration directory.
I tested coc-groovy on a repo without pom or gradle build files and it still connects to the language server resolving local files.

Comment (dansomething) [reply src/classpath.ts]: This was all testing on Linux right? I've been slammed at work, but I'll try to do some testing on Windows this weekend.

Comment (joshzcold) [reply src/classpath.ts]: Yeah all on linux

File: src/classpath.ts
Comment (joshzcold) [reply src/classpath.ts]: Just had a thought on this.
I'm not sure we can guarantee that the error code execution is going to be the error code of the previous command. 

Everything is async and within nodejs so we can't really tell where that error code is coming from unless we specifically checked the pid of the last command 

Better to do what you suggested and check for "BUILD SUCCESS"

Comment (dansomething) [reply src/classpath.ts]: Yeah, I was wondering the same thing. I may be why I checked the message in the first place, but I can't remember. I just updated the gradle-classpath.zip in the comment above. When writing to a file, It will output "BUILD SUCCESS" as the last message if there's no error.

Comment (dansomething) [reply src/classpath.ts]: Will you have time this week to make the update to go back to the check for "BUILD SUCCESS"? I think that's the last bit before it's ready to merge once I do some testing on Windows.

Comment (joshzcold) [reply src/classpath.ts]: Made the change, working for me as expected.

Comment (dansomething) [reply src/classpath.ts]: I've tested on Windows and there are a few minor issues, but I think those we'll be easier to fix and test once these changes are on master.

**Review by dansomething — APPROVED** [review]
It works fine on linux and macos so I think it's ready enough to be merged. I'll followup later with fixes for windows.

**Comment by gVimer** [comment]
> It works fine on linux and macos so I think it's ready enough to be merged. I'll followup later with fixes for windows.

It doesn't work on Windows platform, how to solve it?

**Comment by dansomething** [comment]
@gVimer 
> It doesn't work on Windows platform, how to solve it?

I can't make any guarantees about supporting Windows, but a good place to start would be to open a new Github issue and provide the following information.

**Describe the issue**
A clear and concise description of the problem.

**To Reproduce**
Steps to reproduce the behavior:
1. Go to '...'
2. Open this file '....'
3. Run this command '....'
4. See error

**Expected behavior**
A clear and concise description of what you expected to happen.

**Versions of relevant software**
 - vim/nvim:
 - coc.nvim:
 - coc-groovy:
 - groovy:
 - gradle:
 - maven:

**Relevant logs and error messages**
* View the trace logging output for the Groovy language server:
    * Set `{ "groovy.trace.server": "verbose" }` in `:CocSettings`.
    * Run the command `:CocCommand workspace.showOutput groovy`.
    * Are there any errors?
 * Run `:messages` in vim/nvim. Are any there any relevant errors? 

**Additional context**
Add any other context about the problem here.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Pull request merged**
